classes/ActionScheduler_PastDueMonitor.php

```diff
@@ -0,0 +1,363 @@
+
+		if ( empty( self::$monitor ) ) {
+			$class         = apply_filters( 'action_scheduler_pastdue_actions_monitor_class', 'ActionScheduler_PastDueMonitor' );
+			self::$monitor = new $class();
+		}
+
+		return self::$monitor;
+	}
+
+	/**
+	 * Initialize.
+	 *
+	 * @return void
+	 */
+	public function init() {
+		$this->threshold_seconds       = absint( apply_filters( 'action_scheduler_pastdue_actions_seconds', DAY_IN_SECONDS ) );
+		$this->threshold_minimum       = absint( apply_filters( 'action_scheduler_pastdue_actions_min', 1 ) );
+		$this->interval_check          = absint( apply_filters( 'action_scheduler_pastdue_actions_check_interval', ( $this->threshold_seconds / 4 ) ) );
+		$this->interval_email_seconds  = absint( apply_filters( 'action_scheduler_pastdue_actions_email_interval', HOUR_IN_SECONDS ) );
+		$this->threshold_email_minimum = absint( apply_filters( 'action_scheduler_pastdue_actions_email_min', $this->threshold_minimum ) );
+		$this->notify_methods          = $this->notify_methods();
+
+		$notify_email_to = apply_filters( 'action_scheduler_pastdue_monitor_notify_email_to', get_site_option( 'admin_email' ) );
+
+		if ( ! is_string( $notify_email_to ) || ! is_email( $notify_email_to ) ) {
+			trigger_error( 'Invalid email address provided for past-due actions flooded email notification: using site\'s administrator email address.', E_USER_WARNING );
+			$notify_email_to = get_site_option( 'admin_email' );
+		}
+
+		$this->notify_email_to = $notify_email_to;
+
+		add_action( 'action_scheduler_stored_action', array( $this, 'on_action_stored' ) );
+
+		if ( is_admin() && ( ! defined( 'DOING_AJAX' ) || empty( DOING_AJAX ) ) ) {
+			add_action( 'admin_notices', array( $this, 'action__admin_notices' ) );
+		}
+	}
+
+	/**
+	 * Check if notification method enabled.
+	 *
+	 * @param string $method
+	 * @return bool
+	 */
+	protected function notify( $method ) {
+		return ! empty( $this->notify_methods[ $method ] );
+	}
+
+	/**
+	 * Get enabled notification methods.
+	 *
+	 * @return array<string, bool>
+	 */
+	protected function notify_methods() {
+		$all = array(
+			'admin' => true,
+			'email' => true,
+		);
+
+		$default          = $all;
+		$default['admin'] = function_exists( 'current_user_can' ) && current_user_can( 'manage_options' );
+		$methods          = $default;
+
+		// Apply deprecated filter.
+		if ( has_filter( 'action_scheduler_check_pastdue_actions' ) ) {
+			$methods['admin'] = apply_filters_deprecated(
+				'action_scheduler_check_pastdue_actions',
+				array( $methods['admin'] ),
+				'', // todo: add version number
+				'action_scheduler_pastdue_monitor_notify'
+			);
+		}
+
+		// Allow site devs to specify notification preference.
+		$methods = apply_filters( 'action_scheduler_pastdue_monitor_notify', $methods );
+
+		// Support for scalar and bool values (ex: `__return_false`).
+		if ( is_scalar( $methods ) || is_bool( $methods ) ) {
+			return boolval( $methods ) ? $all : array();
+		}
```

Comment:
If the idea is that the filter yielding `__return_false` would cause an empty array to be returned, this won't work, since `boolval('__return_false')` evaluates to true. Adding the following before this block will solve that, and as a bonus it will add support for filters providing a callable that returns an array:

```php
if ( is_callable( $methods ) ) {
  $methods = $methods();
}
```

Reply:
This code is to support the filter:

```php
add_filter( 'action_scheduler_pastdue_monitor_notify', '__return_false' );
```

Reply:
Yes, I was confused, a snippet like `add_filter( 'action_scheduler_pastdue_monitor_notify', '__return_false' );` will cause `$methods` to become `false` and ths your code would work.

However I've discussed this with my team and we think that adding support for booleans adds unnecessary complexity. Supporting arrays should be enough, client code willing to suppress all the notifications could do just `add_filter( 'action_scheduler_pastdue_monitor_notify', function($methods) { return array(); } );` and you would be able to remove the entire "Support for scalar and bool values" block. Would you mind implementing this change please?

Reply:
You truly think three lines of code is complexity?

Reply:
Oh! Well that's great, because this code is essential functionality: makes it easy for devs to turn off the notifications, regardless of any notifications added later.